src/lib/operators/join_sort_merge.cpp

```diff
@@ -226,6 +260,204 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
   // Represents the result of a value comparison.
   enum class CompareResult { Less, Greater, Equal };
 
+  // In the case of the anti-join, we need to check that there is no combination where all predicates are satisfied. In
+  // the case of the semi-join, we need to check that there is at least one combination where all predicates are
+  // satisfied.
+  void _emit_multipredicate_semi_anti_null_as_false(size_t output_cluster, TableRange left_range,
+                                                    TableRange right_range,
+                                                    MultiPredicateJoinEvaluator& multi_predicate_join_evaluator) {
+    left_range.for_every_row_id(_sorted_left_table, [&](RowID left_row_id) {
+      auto match = false;
+      right_range.for_every_row_id(_sorted_right_table, [&](RowID right_row_id) {
+        if (multi_predicate_join_evaluator.satisfies_all_predicates(left_row_id, right_row_id)) {
+          match = true;
+          return;
+        }
+      });
+      if (match && _mode == JoinMode::Semi) {
+        _emit_left_only(output_cluster, left_row_id);
+      }
+      if (!match && _mode == JoinMode::AntiNullAsFalse) {
+        _emit_left_only(output_cluster, left_row_id);
+      }
+    });
+  }
+
+  void _emit_multipredicate_anti_null_as_true(size_t output_cluster, TableRange left_range, TableRange right_range,
+                                              MultiPredicateJoinEvaluator& multi_predicate_join_evaluator) {
+    left_range.for_every_row_id(_sorted_left_table, [&](RowID left_row_id) {
+      auto match = false;
+      // We need to make sure that there is not one combination between the left rows from the run and the null rows
+      // from the right where all predicates are satisfied.
+      for (const auto& right_row_id : _null_rows_right) {
```

Comment:
Yes, because NULL is always true in this case. Probably it makes more sense to check if there is a match with the right run and then check the null rows (If there was no match). What do you think?

Reply:
Hmm, assuming there are most often less NULLs than values, I think the current approach is fine. It's still baffling me somehow, but multi-predicate joins simply aren't that simple I guess.